### PR TITLE
[FIRRTL][CheckCombCycles] Ignore register self initialization connects

### PR DESCRIPTION
**Pull request by prithayan**
CheckCombCycles should handle register self initialization connects properly. It should not error out on self initialization of individual fields of an aggregate kind of register.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp [hunk 1]
```diff
@@ -301,7 +301,27 @@ class DummySourceNodeIterator
...
+    auto defOp = v.getDefiningOp();
+    return TypeSwitch<Operation *, bool>(defOp)
+        .Case<InstanceOp>([&](InstanceOp instance) { return false; })
+        .Case<SubfieldOp>(
+            [&](SubfieldOp sub) { return isFromRegOp(sub.getInput()); })
+        .Case<SubindexOp>(
+            [&](SubindexOp sub) { return isFromRegOp(sub.getInput()); })
+        .Case<RegOp, RegResetOp>([&](auto reg) { return true; })
+        .Default([&](auto) { return false; });
+  };
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp:324]: You can get this with `isa_and_nonnull<RegOp, RegResetOp>(getFieldRef(v).value)`.

File: lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp [hunk 2]
```diff
@@ -301,7 +301,27 @@ class DummySourceNodeIterator
...
   explicit DummySourceNodeIterator(Node node, bool end = false)
       : node(node), connect(end ? node.context->connects.end()
-                                : node.context->connects.begin()) {}
+                                : node.context->connects.begin()) {
+    // Remove all the register self initialization connects.
+    while (connect != node.context->connects.end() &&
+           ((*connect).getSrc() == (*connect).getDest() &&
+            isFromRegOp((*connect).getDest())))
+      connect++;
+  }
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp:310]: I'm not sure it's accurate. Isn't it possible that register connections appear in the middle of `node.context->connects`?

**Comment by uenoku** [comment]
I think alternatively you can check the subindex/subfield are derived from register at creations. 
https://github.com/llvm/circt/blob/69e674a2efbe1e1c8992a1a50f2542e21cb794cb/lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp#L384

```
 .Case<RegOp, RegResetOp>([&](auto) { 
    return NodeIterator(node, true); 
})
. Case<SubindexOp, SubFieldOp>([&](auto sub) {
   if(isa_and_null<RegOp, RegResetOp>(getFieledRef(sub).value.getDefiningOp()))
    return NodeIterator(node, true);
  return ..;
 });
```

**Comment by prithayan** [comment]
> I think alternatively you can check the subindex/subfield are derived from register at iterator creations.
> 
> https://github.com/llvm/circt/blob/69e674a2efbe1e1c8992a1a50f2542e21cb794cb/lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp#L384
> 
> ```c++
>  .Case<RegOp, RegResetOp>([&](auto) { 
>     return NodeIterator(node, true); 
> })
> . Case<SubindexOp, SubFieldOp>([&](auto sub) {
>    if(isa_and_null<RegOp, RegResetOp>(getFieledRef(sub).value.getDefiningOp()))
>     return NodeIterator(node, true);
>   return ..;
>  });
> ```

You are right, I initially misunderstood the `CombGraphIterator`. We need to update that to handle aggregates in general. Converting this to draft and will spend some more time on this.

**Comment by prithayan** [comment]
@uenoku , implemented using your idea. This should properly handle the registers. But it still does not support aggregates in general, planning to handle it in a followup PR.

File: lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp [hunk 3]
```diff
@@ -358,7 +358,23 @@ class CombGraphIterator
           return InstanceNodeIterator(instance, node, end);
         })
         .Case<SubfieldOp>([&](SubfieldOp subfield) {
-          return SubfieldNodeIterator(subfield, node, end);
+          if (isa_and_nonnull<MemOp>(subfield.getInput().getDefiningOp()))
+            return static_cast<variant_iterator>(
+                SubfieldNodeIterator(subfield, node, end));
+          // This is required to explicitly ignore self initialization connects
+          // for registers.
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp:365]: super nit: loo

File: lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp [hunk 4]
```diff
@@ -358,7 +358,23 @@ class CombGraphIterator
...
+          if (isa_and_nonnull<RegOp, RegResetOp>(
+                  getFieldRefFromValue(subfield).getDefiningOp()))
+            return static_cast<variant_iterator>(NodeIterator(node, true));
+          return static_cast<variant_iterator>(NodeIterator(node, end));
+        })
+        .Case<SubindexOp>([&](SubindexOp sub) {
+          // This is required to explicitly ignore self initialization connects
+          // for registers.
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp:373]: nit: probably "self loops of registers" would be proper word here.

**Review by uenoku — COMMENTED** [review]
LGTM!

**Review by uenoku — APPROVED** [review]

File: lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp (hunk 4 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/CheckCombCycles.cpp]: Updated the comment

**Comment by prithayan** [comment]
Thanks @uenoku , merged the PR.

**Pull request merged**